### PR TITLE
added python version to parse error message

### LINKED ISSUE #3294: Clearer errorr message for "Cannot parse" when targetting multiple version
**Is your feature request related to a problem? Please describe.**

I was working on a project that is only intended to run on Python 3.10. However, the `black` configuration incorrectly listed multiple target versions because I copied the file from a different project.

When I tried to format a file with a match statement, the error was 

> error: cannot format foo.py: Cannot parse: 2:6: match bar:

I was initially confused as I knew that `black` had added Python 3.10 support. Then I noticed my mistake. By saying that the code should target a multiple Python versions, using a match statement would not be valid.

**Describe the solution you'd like**

If `black` encounters a parsing error and it was configured to target multiple Python versions, I think it would be helpful to mention that the syntax "might be valid, but not across all target versions". This would be a hint to the user to think about if they should be using that syntax or that `black` is misconfigured (as it was in my case).

**Describe alternatives you've considered**

It would require more work, but the parser could back track and attempt each of the parser for each of the listed target versions. That would allow `black` provide an even more specific error message. Something like "this is valid for Python 3.10, but targeting Python 3.10 and 3.9", while also being able to identify an invalid syntax.

### PR DESCRIPTION
**Pull request by nnrepos**
### Description
improve parsing error message to make sure users know which version it was run on (fixes #3294).

### Checklist - did you ...
- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -148,6 +148,7 @@
 
 ### Output
 
+- Add python version to parse error message
```
Comment (felix-hilden) [inline CHANGES.md:151]:
```suggestion
- Add python version to parse error message (#3323)
```
Please include the PR number!

**Review by felix-hilden — APPROVED** [review]
Thanks for submitting this, look good apart from a changelog detail 👍

**Review by JelleZijlstra — COMMENTED** [review]
I feel like this will create more user confusion:

```
% black --target-version py36  -c 'a++'
a++
error: cannot format <string>: Cannot parse (python 3.0): 1:3: a++
% black --target-version py39  -c 'a++'
a++
error: cannot format <string>: Cannot parse (python 3.7): 1:3: a++
```
Users are going to wonder why it says "3.0" or "3.7" when they never asked for that version.

**Comment by nnrepos** [comment]
> I feel like this will create more user confusion:
> 
> ```
> % black --target-version py36  -c 'a++'
> a++
> error: cannot format <string>: Cannot parse (python 3.0): 1:3: a++
> % black --target-version py39  -c 'a++'
> a++
> error: cannot format <string>: Cannot parse (python 3.7): 1:3: a++
> ```
> 
> Users are going to wonder why it says "3.0" or "3.7" when they never asked for that version.

ok, can you help me understand why we parse `python 3.0`, when a user asks to parse `py36`?

**Comment by JelleZijlstra** [comment]
> ok, can you help me understand why we parse `python 3.0`, when a user asks to parse `py36`?

We don't have separate grammars for every version. For py36 we use a grammar with version set to (3, 0).

**Comment by nnrepos** [comment]
> > ok, can you help me understand why we parse `python 3.0`, when a user asks to parse `py36`?
> 
> We don't have separate grammars for every version. For py36 we use a grammar with version set to (3, 0).

ok, so do we have an access to the version that the user provided somewhere?

**Comment by JelleZijlstra** [comment]
It's the target version, which gets passed as a command-line parameter and thence as an argument to `main()`.

**Comment by nnrepos** [comment]
hey @JelleZijlstra @felix-hilden, i've made some changes, please let me know if this is helpful 😄

**Comment by JelleZijlstra** [comment]
I'm still not enthusiastic about this:
- Hardcoding the list of versions is a little hacky and hard to maintain
- I'm afraid the error message may still be confusing for users, but don't have a better suggestion

**Comment by nnrepos** [comment]
> I'm still not enthusiastic about this:
> 
> * Hardcoding the list of versions is a little hacky and hard to maintain
> * I'm afraid the error message may still be confusing for users, but don't have a better suggestion

* i've used the same hard-coding that is used by the parser, since there weren't any constants/functions which can be used to do that. that can be refactored of course.
* i think users are smart enough to understand what a range of numbers is, we're all programmers here...

feel free to close this if you're not interested

**Pull request closed without merging**